File: common/rdr/TLSInStream.cxx
/* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 * Copyright (C) 2005 Martin Koegler
 * Copyright (C) 2010 TigerVNC Team
 * Copyright (C) 2012-2021 Pierre Ossman for Cendio AB
 * 
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <core/Exception.h>
#include <core/LogWriter.h>

#include <rdr/TLSException.h>
#include <rdr/TLSInStream.h>

#include <errno.h>

#ifdef HAVE_GNUTLS 
using namespace rdr;

static core::LogWriter vlog("TLSInStream");

ssize_t TLSInStream::pull(gnutls_transport_ptr_t str, void* data, size_t size)
{
  TLSInStream* self= (TLSInStream*) str;
  InStream *in = self->in;

  self->streamEmpty = false;
  self->saved_exception = nullptr;

  try {
    if (!in->hasData(1)) {
      self->streamEmpty = true;
      gnutls_transport_set_errno(self->session, EAGAIN);
      return -1;
    }

    if (in->avail() < size)
      size = in->avail();
  
    in->readBytes((uint8_t*)data, size);
  } catch (end_of_stream&) {
    return 0;
  } catch (std::exception& e) {
    core::socket_error* se;
    vlog.error("Failure reading TLS data: %s", e.what());
    se = dynamic_cast<core::socket_error*>(&e);
    if (se)
      gnutls_transport_set_errno(self->session, se->err);
    else
      gnutls_transport_set_errno(self->session, EINVAL);
    self->saved_exception = std::current_exception();
    return -1;
  }

  return size;
}

TLSInStream::TLSInStream(InStream* _in, gnutls_session_t _session)
  : session(_session), in(_in)
{
  gnutls_transport_ptr_t recv, send;

  gnutls_transport_set_pull_function(session, pull);
  gnutls_transport_get_ptr2(session, &recv, &send);
  gnutls_transport_set_ptr2(session, this, send);
}

TLSInStream::~TLSInStream()
{
  gnutls_transport_set_pull_function(session, nullptr);
}

bool TLSInStream::fillBuffer()
{
  size_t n = readTLS((uint8_t*) end, availSpace());
  if (n == 0)
    return false;
  end += n;

  return true;
}

size_t TLSInStream::readTLS(uint8_t* buf, size_t len)
{
  int n;

  while (true) {
    streamEmpty = false;
    n = gnutls_record_recv(session, (void *) buf, len);
    if (n == GNUTLS_E_INTERRUPTED || n == GNUTLS_E_AGAIN) {
      // GnuTLS returns GNUTLS_E_AGAIN for a bunch of other scenarios
      // other than the pull function returning EAGAIN, so we have to
      // double check that the underlying stream really is empty
      if (!streamEmpty)
        continue;
      else
        return 0;
    }
    break;
  };

  if (n == GNUTLS_E_PULL_ERROR)
    std::rethrow_exception(saved_exception);

  if (n < 0)
    throw tls_error("readTLS", n);

  if (n == 0)
    throw end_of_stream();

  return n;
}

#endif

File: java/com/tigervnc/rdr/TLSInStream.java
/* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 * Copyright (C) 2005 Martin Koegler
 * Copyright (C) 2010 TigerVNC Team
 * Copyright (C) 2011-2019 Brian P. Hinz
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
 * USA.
 */

package com.tigervnc.rdr;

import java.nio.ByteBuffer;
import java.nio.channels.*;
import javax.net.ssl.*;

import com.tigervnc.network.*;

public class TLSInStream extends InStream {

  static final int defaultBufSize = 16384;

  public TLSInStream(InStream _in, SSLEngineManager _manager) {
    in = (FdInStream)_in;
    manager = _manager;
    offset = 0;
    SSLSession session = manager.getSession();
    bufSize = session.getApplicationBufferSize();
    b = new byte[bufSize];
    ptr = end = start = 0;
  }

  public final int pos() {
    return offset + ptr - start;
  }

  public final void startTiming() {
    in.startTiming();
  }

  public final void stopTiming() {
    in.stopTiming();
  }

  public final long kbitsPerSecond() {
    return in.kbitsPerSecond();
  }

  public final long timeWaited() {
    return in.timeWaited();
  }

  protected final int overrun(int itemSize, int nItems, boolean wait) {
    if (itemSize > bufSize)
      throw new Exception("TLSInStream overrun: max itemSize exceeded");

    if (end - ptr != 0)
      System.arraycopy(b, ptr, b, 0, end - ptr);

    offset += ptr - start;
    end -= ptr - start;
    ptr = start;

    while ((end - start) < itemSize) {
      int n = readTLS(b, end, start + bufSize - end, wait);
      if (!wait && n == 0)
        return 0;
      end += n;
    }

    int nAvail;
    nAvail = (end - ptr) / itemSize;
    if (nAvail < nItems)
      return nAvail;

    return nItems;
  }

  protected int readTLS(byte[] buf, int bufPtr, int len, boolean wait)
  {
    int n = -1;

    try {
      n = manager.read(ByteBuffer.wrap(buf, bufPtr, len), len);
    } catch (java.io.IOException e) {
      e.printStackTrace();
    }

    if (n < 0) throw new TLSException("readTLS", n);

    return n;
  }

  private SSLEngineManager manager;
  private int offset;
  private int start;
  private int bufSize;
  private FdInStream in;
}
